### PR TITLE
usb: stm32: add usb_dc_ep_read_wait/conitinue

### PR DESCRIPTION
**Pull request by qianfan-Zhao**
usb mass example need usb_dc_ep_read_wait/continue API. but stm32's usb device controller driver doesn't define. so i added thats and test usb mass storage with RAM DISK on stm32f4 series successful.

### CODE REVIEW

**Comment by zhaozhongchen** [comment]
@andyross @erwango @galak @superna9999 @loicpoulain Could you review my code?

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -511,18 +511,39 @@ int usb_dc_ep_read(const u8_t ep, u8_t *const data, const u32_t max_data_len,
...
-		*read_bytes = read_count;
-	}
+	return 0;
+}
+
+int usb_dc_ep_read(const u8_t ep, u8_t *const data, const u32_t max_data_len,
+		   u32_t * const read_bytes)
+{
+	usb_dc_ep_read_wait(ep, data, max_data_len, read_bytes);
+	usb_dc_ep_read_continue(ep);
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c]: Could you please check returned values. Now it's always 0 (for now).

**Review by loicpoulain — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_stm32.c (hunk 1 above)
Comment (zhaozhongchen) [reply drivers/usb/device/usb_dc_stm32.c]: the previous version always return 0, so there i haven't change it.
you means there need return negative when fails, like usb_dc_dw.c did?

Comment (loicpoulain) [reply drivers/usb/device/usb_dc_stm32.c]: I mean usb_dc_ep_read_wait and usb_dc_ep_read_continue method return value we should not ignore.

Comment (zhaozhongchen) [reply drivers/usb/device/usb_dc_stm32.c]: I had fix this, please review again.

**Comment by zhaozhongchen** [comment]
@ydamigos  coulnd you review this branch?

**Review by ydamigos — CHANGES_REQUESTED** [review]
> author email (QianFan <qianfanguijin@163.com>) does not follow the syntax: First Last <email>.

Please fix the issue above and squash the two commits in one.

File: drivers/usb/device/usb_dc_stm32.c [hunk 2]
```diff
@@ -497,8 +497,24 @@ int usb_dc_ep_read_wait(u8_t ep, u8_t *data, u32_t max_data_len,
...
+
+	if (!data && max_data_len) {
+		SYS_LOG_ERR("Wrong arguments");
+		return -EINVAL;
+	}
+
+	if (!data && !max_data_len) {
+		/* When both buffer and max data to read are zero return
+		 * the avaliable data in buffer
+		 */
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c]: keep the 'goto done' here to have unified exit.

**Comment by zhaozhongchen** [comment]
@ydamigos @loicpoulain I had fix the issue and squash all commit in one. please review the code.

File: drivers/usb/device/usb_dc_stm32.c [hunk 3]
```diff
@@ -488,16 +488,29 @@ int usb_dc_ep_write(const u8_t ep, const u8_t *const data,
...
+	}
+
+	if (!data && max_data_len) {
+		SYS_LOG_ERR("Wrong arguments");
+		return -EINVAL;
+	}
+
+	if (!data && !max_data_len) {
+		/* When both buffer and max data to read are zero return
+		 * the avaliable data in buffer
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c]: typo: available

**Comment by zhaozhongchen** [comment]
@loicpoulain I had fix the typo, do you have any suggests?

**Review by loicpoulain — APPROVED** [review]
LGTM

File: drivers/usb/device/usb_dc_stm32.c [hunk 4]
```diff
@@ -488,16 +488,29 @@ int usb_dc_ep_write(const u8_t ep, const u8_t *const data,
...
+		SYS_LOG_ERR("Wrong endpoint direction: 0x%02x", ep);
+		return -EINVAL;
+	}
+
+	if (!data && max_data_len) {
+		SYS_LOG_ERR("Wrong arguments");
+		return -EINVAL;
+	}
+
+	if (!data && !max_data_len) {
```
Comment (ydamigos) [inline drivers/usb/device/usb_dc_stm32.c]: Could we refactor the code to be more clear. E.g.:
```
if (!EP_IS_OUT(ep)) { ... }

if (!max_data_len) { ... }

...

if (data && read_count) {...}
```

Comment (zhaozhongchen) [reply drivers/usb/device/usb_dc_stm32.c]: seems there is no differences

Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm32.c]: We have one less if condition and the code is easier to read at least for me.

**Review by loicpoulain — COMMENTED** [review]
I've integrated this patch with minor rework in PR #5214

**Comment by ydamigos** [comment]
>  I've integrated this patch with minor rework in PR #5214

PR #5214 was merged.

**Pull request closed without merging**